Add tests for api endpoints: `timeline-event`, `search`

kienstra opened the pull request:
Closes part of #28598

### Description

Adds unit tests for these endpoints:

* `/search/models` (GET)
* `/timeline-event/:id` (DELETE)

### How to verify

```
clojure -X:dev:test :only metabase.api.search-test metabase.api.timeline-event-test
```

### Demo

![tests-passed](https://github.com/kienstra/metabase/assets/4063887/bb06267a-9867-4960-8dc4-0e749674d22c)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

kienstra committed:
Add a unit test for /api/pulse/preview_card_info/:id

For the GET endpoint
/api/pulse/preview_card_info/:id

kienstra committed:
Add a unit test for the GET endpoint /search/models

kienstra committed:
Add a unit test for DELETE /api/timeline-event/:id

kienstra committed:
Add a unit test for DELETE /api/timeline-event/:id

kienstra committed:
Merge branch 'master' into add/api-unit-tests

kienstra commented:
Hi @qnkhuat,
Could you use this PR for more tests from your #28598?

[Here's a PR](https://github.com/kienstra/metabase/pull/2) with GItHub actions passing.

Thanks for your review of #33776 and for your patience.

Hope you had a great weekend.

qnkhuat commented on test/metabase/api/pulse_test.clj:
```diff
@@ -1103,6 +1103,23 @@
                             s/Keyword s/Any}
                      body)))))))))
 
+(deftest preview-card-info-pulse-test
```
in general simply testing the shape of the response is not enough, but I think in this case it's okay because this endpoint is not used anymore and will be removed as part of https://github.com/metabase/metabase/issues/34147

qnkhuat replied:
some example of what to test comes into mind:
- the col_count and row_count should matches with executing the card
- matching regex of `pulse_card_html` to contains certain key element
- :pulse_card_name,  :pulse_card_type should be expected to equal some values

These are just some suggestions for future tests, you don't need to add for this one as it'll be removed soon.

kienstra replied:
Good ideas, like testing that `pulse_card_html` has a "key element."

If it's alright, I'll leave this commit as it, to be deleted in https://github.com/metabase/metabase/issues/34147.

kienstra replied:
Actually, I removed  `(deftest preview-card-info-pulse-test)` in https://github.com/metabase/metabase/pull/34072/commits/6c2b4d41442ddda3726098874e7e0af5006c7e7d if that's OK.

qnkhuat commented on test/metabase/api/timeline_event_test.clj:
```diff
@@ -66,3 +66,20 @@
+                                             :timeline_id (u/the-id timeline)}]
+      (let [event-id (u/the-id event)]
+        (testing "delete the timeline-event by `id`"
+          (is (= "Example Event"
```
in general, if we want to test the existence of sth we can just fetch it from db directly, like `(t2/select-one-fn :name :model/TimelineEvent :id event-id)`.

But in this case I think we don't need to do that because we explicitly created it with the with-temp above. so this could be removed

kienstra replied:
Good idea, it was created in `with-temp`, so no need to test it.

Remove in https://github.com/metabase/metabase/pull/34072/commits/851ad805eadc347e91088f7e378c7f634c9d9c4d.

qnkhuat commented on test/metabase/api/timeline_event_test.clj:
```diff
@@ -66,3 +66,20 @@
+          (is (= "Example Event"
+                 (->> (mt/user-http-request :rasta :get 200 (str "timeline-event/" event-id))
+                      :name)))
+          (is (= nil
+                 (mt/user-http-request :rasta :delete 204 (str "timeline-event/" event-id))))
```
can you add another test to check the response when it deletes a timeline event that doesn't exist?

kienstra replied:
Yes, good idea.

qnkhuat commented on test/metabase/api/timeline_event_test.clj:
```diff
@@ -66,3 +66,20 @@
+    (mt/with-temp [Collection    collection {:name "Example Data"}
+                   Timeline      timeline   {:name          "Some Events"
+                                             :collection_id (u/the-id collection)}
+                   TimelineEvent event      {:name        "Example Event"
```
```suggestion
                   TimelineEvent {event-id :id}      {:name        "Example Event"
```

we only care about the event-id as part of the test, so might as well destructure it so we don't have to do `(u/the-id event)` later.

qnkhuat replied:
same with timeline-id

kienstra replied:
Thanks, [committed here](https://github.com/metabase/metabase/compare/76aa237563cfaeec909e00b4f8a903d9c5be51f0...f7f494ec2164fd8f006a7f752dc72279fa872238?expand=1).

…with your suggestion applied for `timeline-id` and `collection-id`.

qnkhuat reviewed (changes requested):
Thanks for doing this. See my comments for feedback

kienstra committed:
Commit Ngoc's suggestion: Update test/metabase/api/timeline_event_test.clj

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

kienstra committed:
Apply Ngoc's suggestion to destructure :timeline_id

kienstra committed:
Remove test for event that I created in with-temp

kienstra committed:
Fix failed test I introduced, change collection to collection-id

kienstra committed:
Test deleting a non-existent timeline event by id

kienstra committed:
Move my search-models-test lower

kienstra committed:
Copy-paste Ngoc's query-model-set-test and macro

https://github.com/metabase/metabase/pull/32624/files#diff-734923de9c62f49d6b708d64a214ee38d7d0fb1b2c20343bff7b947a6021b021R297-R332

https://github.com/metabase/metabase/pull/32624/files#diff-54344d8b75319d40a4cb2a2afe935500c23d34f0f3c63c51f005493d4dc4f8f1R690-R732

kienstra commented on test/metabase/test/util.clj:
```diff
@@ -681,6 +687,49 @@
+  (with-model-cleanup [:model/ModerationReview]
+    (doseq [card-or-id card-or-ids]
+      (doseq [status ["verified" nil "verified"]]
+        ;; create multiple moderation review for a card, but the end result is it's still verified
+        (moderation-review/create-review!
+         {:moderated_item_id   (u/the-id card-or-id)
+          :moderated_item_type "card"
+          :moderator_id        ((requiring-resolve 'metabase.test.data.users/user->id) :rasta)
+          :status              status})))
+    (thunk)))
+
+(defmacro with-verified-cards
+  "Execute the body with all `card-or-ids` verified."
+  [card-or-ids & body]
+  `(do-with-verified-cards ~card-or-ids (fn [] ~@body)))
+
+(deftest with-verified-cards-test
+  (t2.with-temp/with-temp
+    [:model/Card {card-id :id} {}]
+    (with-verified-cards [card-id]
+      (is (=? #{{:moderated_item_id   card-id
+                 :moderated_item_type :card
+                 :most_recent         true
+                 :status              "verified"}
+                {:moderated_item_id   card-id
+                 :moderated_item_type :card
+                 :most_recent         false
+                 :status              nil}
+                {:moderated_item_id   card-id
+                 :moderated_item_type :card
+                 :most_recent         false
+                 :status              "verified"}}
+              (t2/select-fn-set #(select-keys % [:moderated_item_id :moderated_item_type :most_recent :status])
+                                :model/ModerationReview
+                                :moderated_item_id card-id
+                                :moderated_item_type "card"))))
+    (testing "everything is cleaned up after the macro"
+      (is (= 0 (t2/count :model/ModerationReview
+                         :moderated_item_id card-id
+                         :moderated_item_type "card"))))))
```
Thanks to Ngoc [for this](https://github.com/metabase/metabase/pull/32624/files#diff-54344d8b75319d40a4cb2a2afe935500c23d34f0f3c63c51f005493d4dc4f8f1R690-R732).

kienstra replied:
Reverted this, as it's for a test that's not in scope.

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -755,3 +792,31 @@
+                 (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term :models "card"))))
+
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card" "table" "database"}
+                 (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term :models "card" :models "dashboard")))))))))
```
[Thanks to Ngoc](https://github.com/metabase/metabase/pull/32624/files#diff-734923de9c62f49d6b708d64a214ee38d7d0fb1b2c20343bff7b947a6021b021) for most of these search tests

kienstra committed:
Remove "database" and "table" from test expectation

Those failed in this PR,
though they passed in 32624

kienstra committed:
Revert test for pulse/preview_card_info/, as it'll be removed

kienstra committed:
Remove with-verified-cards, which I needlessly copied into this PR

That's needed in 32624,
but not this PR.

kienstra committed:
Add a test for the archived-string param

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -285,6 +285,16 @@
     (with-search-items-in-root-collection "test"
       (is (= [] (:available_models (mt/user-http-request :crowberto :get 200 "search?q=noresults")))))))
 
+(deftest query-model-set-test
+  (let [search-term "query-model-set"]
+    (with-search-items-in-root-collection search-term
+      (testing "should return a list of models that search result will return"
+        (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term)))))
+      (testing "should not return models when there is no search result"
+        (is (= #{}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :q "noresults"))))))))
+
```
Copy-pasted [from Ngoc's PR](https://github.com/metabase/metabase/pull/32624/files#diff-734923de9c62f49d6b708d64a214ee38d7d0fb1b2c20343bff7b947a6021b021R297-R302).

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -755,3 +765,55 @@
       (mt/user-http-request :crowberto :get 200 "search" :q "test" :archived true)
       (is (empty? (snowplow-test/pop-event-data-and-user-id!))))))
+
+(deftest available-models-should-be-independent-of-models-param-test
+  (testing "if a search request includes `models` params, the `available_models` from the response should not be restricted by it"
+    (let [search-term "Available models"]
+      (with-search-items-in-root-collection search-term
+        (testing "GET /api/search"
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+                 (-> (mt/user-http-request :crowberto :get 200 "search" :q search-term :models "card")
+                     :available_models
+                     set)))
+
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+                 (-> (mt/user-http-request :crowberto :get 200 "search" :q search-term :models "card" :models "dashboard")
+                     :available_models
+                     set))))
+
+        (testing "GET /api/search/models"
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+                 (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term :models "card"))))
+
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+                 (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term :models "card" :models "dashboard")))))))))
```
Copy-pasted [from Ngoc's PR](https://github.com/metabase/metabase/pull/32624/files#diff-734923de9c62f49d6b708d64a214ee38d7d0fb1b2c20343bff7b947a6021b021R1029-R1049).

Also, I [removed](https://github.com/kienstra/metabase/pull/2/commits/66bd9695f4e6b89eb8f84f38c1da33b9a6f722da#) `"table"` and `"database"` from the expected values to get the tests to pass in this PR. Of course, they passed in #32624.

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -755,3 +765,55 @@
+          (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+                 (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term :models "card" :models "dashboard")))))))))
+
+(deftest models-table-db-id-test
+  (testing "search/models request includes `table-db-id` param"
+    (with-search-items-in-root-collection "Available models"
+      (testing "`table-db-id` is invalid"
+        (is (=? {:errors {:table-db-id "nullable value must be an integer greater than zero."}}
+                (mt/user-http-request :crowberto :get 400 "search/models" :table-db-id -1))))
+      (testing "`table-db-id` is for a non-existent database"
+        (is (= #{"dashboard" "database" "segment" "collection" "action" "metric"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :table-db-id Integer/MAX_VALUE)))))
+      (testing "`table-db-id` is for an existing database"
+        (is (= #{"dashboard" "database" "segment" "collection" "action" "metric" "card" "dataset" "table"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :table-db-id (mt/id)))))))))
+
+(deftest models-archived-string-test
+  (testing "search/models request includes `archived-string` param"
+    (with-search-items-in-root-collection "Available models"
+      (mt/with-temp [Card        {model-id :id} action-model-params
+                     Action      _              (archived {:name     "test action"
+                                                           :type     :query
+                                                           :model_id model-id})]
+      (testing "`archived-string` is invalid"
+        (is (=? {:message "Invalid input: [\"value must be a valid boolean string ('true' or 'false').\"]"}
+                (mt/user-http-request :crowberto :get 500 "search/models" :archived-string 1))))
+      (testing "`archived-string` is 'false'"
+        (is (= #{"dashboard" "database" "segment" "collection" "action" "metric" "card" "dataset" "table"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :archived-string "false")))))
+      (testing "`archived-string` is 'true'"
+        (is (= #{"action"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :archived-string "true")))))))))
```
Tests the [remaining 2 parameters](https://github.com/metabase/metabase/blob/6927d396b661f0cd4c3f427257a43aa66f5ce37f/docs/api/search.md#get-apisearchmodels) to `/api/search/models`, now that the `q` parameter is tested in `(deftest query-model-set-test`

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -285,6 +285,16 @@
     (with-search-items-in-root-collection "test"
       (is (= [] (:available_models (mt/user-http-request :crowberto :get 200 "search?q=noresults")))))))
 
+(deftest query-model-set-test
+  (let [search-term "query-model-set"]
+    (with-search-items-in-root-collection search-term
+      (testing "should return a list of models that search result will return"
+        (is (= #{"dashboard" "dataset" "segment" "collection" "action" "metric" "card"}
+               (set (mt/user-http-request :crowberto :get 200 "search/models" :q search-term)))))
```
Copy-pasted [from Ngoc's PR](https://github.com/metabase/metabase/pull/32624/files#diff-734923de9c62f49d6b708d64a214ee38d7d0fb1b2c20343bff7b947a6021b021R297-R302).

kienstra commented on test/metabase/api/search_test.clj:
```diff
@@ -737,6 +737,12 @@
         ;; into keep this number as low as we can
         (is (= 16 (call-count)))))))
 
+(deftest search-models-test
```
Good idea, how do the tests for `search_test.clj` look now?

I copy-pasted 2 of your tests, and added some of mine.

kienstra commented:
Hi @qnkhuat,
Thanks for the care you put into reviewing this, and for [your](https://github.com/metabase/metabase/pull/34072#pullrequestreview-1650001838) [tests](https://github.com/metabase/metabase/pull/34072#discussion_r1340862892) that I copy-pasted.

Your suggestions should all be applied now.

kienstra committed:
Merge branch 'master' into add/api-unit-tests

kienstra committed:
Merge branch 'master' into add/api-unit-tests

kienstra committed:
Merge branch 'master' into add/api-unit-tests

qnkhuat commented:
Sorry it took so longs to review. Next time please help me request for another review!

I created this [PR](https://github.com/metabase/metabase/pull/34479) to run CI, once it's green this should be good to go.

kienstra commented:
Hi @qnkhuat,
Oh all good! Sorry for not requesting another review.

qnkhuat reviewed (approved):
This LGTM and CI is green https://github.com/metabase/metabase/pull/34479

kienstra commented:
@qnkhuat,
Thanks a lot for your [code](https://github.com/metabase/metabase/pull/34072#discussion_r1340862892) and [suggestions](https://github.com/metabase/metabase/pull/34072#event-10588860609) for this PR!

Hope you have a great week.
